.github/workflows/iroha2-dev-pr.yml

```diff
@@ -144,44 +150,67 @@ jobs:
+          curl -sSL https://install.python-poetry.org | python3 -
+          echo 'export PATH="${{ env.POETRY_PATH }}:$PATH"' >> /etc/profile
+          source /etc/profile
+      - name: Install Torii API Dependencies
         working-directory: torii/pytests
+        run: ${{ env.POETRY_PATH }} install
+      - name: Build and Tag Docker Images
+        uses: docker/build-push-action@v6
+        if: always()
+        with:
+          context: .
+          load: true
+          file: Dockerfile
+          tags: |
+            hyperledger/iroha:local
+            hyperledger/iroha:dev
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+      - name: Test docker-compose.single.yml
+        run: |
+          docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.single.yml up --wait || exit 1
+          docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.single.yml down
+      - name: Test docker-compose.local.yml
         run: |
-          poetry install
-      - name: Run torii api tests
+          docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.local.yml up --wait || exit 1
+          docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.local.yml down
+      - name: Run docker-compose.yml containers
+        run: docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.yml up --wait || exit 1
+      - name: Run Torii API Tests
         working-directory: torii/pytests
+        run: ${{ env.POETRY_PATH }} run pytest
+      - name: Copy Client CLI Binary from Iroha Container
         run: |
-          poetry run pytest
-      - name: Install client cli dependencies using Poetry
+          mkdir -p ${{ env.TEST_DIR }}
+          docker cp ${{ env.IROHA_CONTAINER }}:/usr/local/bin/${{ env.IROHA_BIN }} ${{ env.TEST_DIR }}
+          cp ./defaults/client.toml ${{ env.TEST_DIR }}
+      - name: Make Binaries Executable
+        run: chmod +x ${{ env.TEST_DIR }}/${{ env.IROHA_BIN }}
+      - name: Install Client CLI Dependencies
         working-directory: client_cli/pytests
-        run: |
-          poetry install
-      - name: Run client cli tests
+        run: ${{ env.POETRY_PATH }} install
+      - name: Run Client CLI Tests
         working-directory: client_cli/pytests
         env:
-          # prepared by `test_env.py`
-          CLIENT_CLI_BINARY: ../../test/iroha
-          CLIENT_CLI_CONFIG: ../../test/client.toml
-        run: |
-          poetry run pytest
-      - name: Cleanup test environment
-        run: |
-          ./scripts/test_env.py cleanup
+          CLIENT_CLI_BINARY: ../../${{ env.TEST_DIR }}/${{ env.IROHA_BIN }}
+          CLIENT_CLI_CONFIG: ../../${{ env.TEST_DIR }}/client.toml
+        run: ${{ env.POETRY_PATH }} run pytest
+      - name: Wipe docker-compose.yml containers
+        run: docker compose -f ${{ env.DOCKER_COMPOSE_PATH }}/docker-compose.yml down
```

Comment:
Why? A good practice is to stop & remove containers after all tests are done even on the (_currently_) spot runners

Reply:
IMO you should tear down execution context regardless of whether your tests succeeded or not

Reply:
If some tests are failed, this workflow will be failed as usual and how it's going now. We just wipe testing containers at the end if everything before was passed successfully

Reply:
I understand what this does. But will this step be executed if the previous fails? Don't you want to stop docker containers regardless of what has happened in the previous steps

Reply:
I've added the condition to remove containers if the latest `client_cli` test was failed. But it will not always work if any of previous steps was failed since we should not allow all job steps execution and Actions can not manage this process smarter